worker/download.go

```diff
@@ -0,0 +1,1088 @@
+		mgr.mu.Unlock()
+	}
+
+	// calculate the offset and length
+	offset, length := slice.SectorRegion()
+
+	// build sector info
+	hostToSectors := make(map[types.PublicKey][]sectorInfo)
+	for sI, s := range slice.Shards {
+		hostToSectors[s.Host] = append(hostToSectors[s.Host], sectorInfo{s, sI})
+	}
+
+	// create slab download
+	return &slabDownload{
+		mgr: mgr,
+
+		dID:       dID,
+		sID:       sID,
+		created:   time.Now(),
+		index:     slabIndex,
+		minShards: int(slice.MinShards),
+		offset:    offset,
+		length:    length,
+
+		hostToSectors: hostToSectors,
+		used:          make(map[types.PublicKey]struct{}),
+
+		sectors: make([][]byte, len(slice.Shards)),
+	}, finishFn
+}
+
+func (mgr *downloadManager) ongoingDownloads() int {
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+	return len(mgr.ongoing)
+}
+
+func (mgr *downloadManager) downloadSlab(ctx context.Context, dID id, slice object.SlabSlice, index int, responseChan chan *slabDownloadResponse, nextSlabChan chan struct{}) {
+	// add tracing
+	ctx, span := tracing.Tracer.Start(ctx, "downloadSlab")
+	defer span.End()
+
+	// prepare the download
+	slab, finishFn := mgr.newSlabDownload(ctx, dID, slice, index)
+	defer finishFn()
+
+	// download shards
+	resp := &slabDownloadResponse{index: index}
+	resp.shards, resp.err = slab.downloadShards(ctx, nextSlabChan)
+
+	// send the response
+	select {
+	case <-ctx.Done():
+	case responseChan <- resp:
+	}
+}
+
+func (d *downloader) stats() downloaderStats {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	return downloaderStats{
+		avgSpeedMBPS: d.statsDownloadSpeedBytesPerMS.Average() * 0.008,
+		healthy:      d.consecutiveFailures == 0,
+		numDownloads: d.numDownloads,
+	}
+}
+
+func (d *downloader) isStopped() bool {
+	select {
+	case <-d.stopChan:
+		return true
+	default:
+	}
+	return false
+}
+
+func (d *downloader) processQueue(hp hostProvider) {
+outer:
+	for {
+		// wait for work
+		select {
+		case <-d.signalWorkChan:
+		case <-d.stopChan:
+			return
+		}
+
+		for {
+			// check if we are stopped
+			if d.isStopped() {
+				return
+			}
+
+			// fill batch
+			var batch []*sectorDownloadReq
+			for len(batch) < maxConcurrentSectorsPerHost {
+				if req := d.pop(); req == nil {
+					break
+				} else if req.done() {
+					continue
+				} else {
+					batch = append(batch, req)
+				}
+			}
+			if len(batch) == 0 {
+				continue outer
+			}
+
+			fmt.Printf("DEBUG PJ: downloader %v is processing batch of size %v\n", d.host.HostKey(), len(batch))
+
+			var mu sync.Mutex
+			var completed int
+			var failed int
+			var first time.Time
+
+			// launch workers
+			var wg sync.WaitGroup
+			jobsChan := make(chan *sectorDownloadReq)
+			for i := 0; i < len(batch); i++ {
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					for job := range jobsChan {
+						// check if we are stopped
+						if d.isStopped() {
+							return
+						}
+
+						// execute the job
+						start := time.Now()
+						err := d.execute(job)
+						d.trackFailure(err)
```

Comment:
Keep in mind that not every failure might warrant punishing the host. e.g. a sector might not exist. That sucks for this download but doesn't necessarily affect any other download.

Migrations might hit that occasionally and might end up ruining the performance for regular uploads as well by tracking it.